src/d3d9/d3d9_initializer.cpp

```diff
@@ -113,41 +113,35 @@ namespace dxvk {
     // If the buffer is mapped, we can write data directly
     // to the mapped memory region instead of doing it on
     // the GPU. Same goes for zero-initialization.
-    const D3D9_COMMON_TEXTURE_DESC* desc = pTexture->Desc();
-    for (uint32_t a = 0; a < desc->ArraySize; a++) {
-      for (uint32_t m = 0; m < desc->MipLevels; m++) {
-        uint32_t subresource = pTexture->CalcSubresource(a, m);
-        void* mapPtr = pTexture->GetData(subresource);
-        uint32_t length = pTexture->GetMipSize(subresource);
-
-        if (pInitialData != nullptr) {
-          VkExtent3D mipExtent = pTexture->GetExtentMip(m);
-          const DxvkFormatInfo* formatInfo = lookupFormatInfo(pTexture->GetFormatMapping().FormatColor);
-          VkExtent3D blockCount = util::computeBlockCount(mipExtent, formatInfo->blockSize);
-          uint32_t pitch = blockCount.width * formatInfo->elementSize;
-          uint32_t alignedPitch = align(pitch, 4);
-
-          util::packImageData(
-            mapPtr,
-            pInitialData,
-            pitch,
-            pitch * blockCount.height,
-            alignedPitch,
-            alignedPitch * blockCount.height,
-            D3D9CommonTexture::GetImageTypeFromResourceType(pTexture->GetType()),
-            mipExtent,
-            pTexture->Desc()->ArraySize,
-            formatInfo,
-            VK_IMAGE_ASPECT_COLOR_BIT);
-        } else {
-          std::memset(
-            mapPtr, 0,
-            length);
-        }
-      }
+    void* mapPtr = pTexture->GetData(0);
+    if (pInitialData) {
+      // Initial data is only supported for textures with 1 subresource
+      VkExtent3D mipExtent = pTexture->GetExtentMip(0);
+      const DxvkFormatInfo* formatInfo = lookupFormatInfo(pTexture->GetFormatMapping().FormatColor);
+      VkExtent3D blockCount = util::computeBlockCount(mipExtent, formatInfo->blockSize);
+      uint32_t pitch = blockCount.width * formatInfo->elementSize;
+      uint32_t alignedPitch = align(pitch, 4);
+
+      util::packImageData(
+        mapPtr,
+        pInitialData,
+        pitch,
+        pitch * blockCount.height,
+        alignedPitch,
+        alignedPitch * blockCount.height,
+        D3D9CommonTexture::GetImageTypeFromResourceType(pTexture->GetType()),
+        mipExtent,
+        pTexture->Desc()->ArraySize,
+        formatInfo,
+        VK_IMAGE_ASPECT_COLOR_BIT);
+    } else {
+      // All subresources are allocated in one chunk of memory.
+      // So we can just get the pointer for subresource 0 and memset all of them at once.
+      std::memset(
+        mapPtr, 0,
+        pTexture->GetTotalSize());
     }
-    if (pTexture->GetMapMode() == D3D9_COMMON_TEXTURE_MAP_MODE_UNMAPPABLE)
-      pTexture->UnmapData();
+    pTexture->UnmapData();
```

Comment:
Looks like the logic here changed and now it always calls UnmapData instead of just when == D3D9_COMMON_TEXTURE_MAP_MODE_UNMAPPABLE

Reply:
Yeah, that's fine. Unmap will check if it's mapped and just exit out.

Reply:
Sorry I don't see how that is equivalent.

Reply:
If the map mode isn't UNMAPPABLE, the memory is zero initialized. In that case all methods of m_memory exit out without doing anything.